### ISSUE TITLE
adding an augmentation 

### ISSUE THREAD

**Issue by natamendelson** [MIC-DKFZ/nnUNet#2176]
hi, suppose i want to add a specific augmentation to the dataset while training, where is the right place to do it?
i noticed that you have a function get_training_transforms in nnUNetTrainer.py and i tried to add my own transformation there but it appears that the input array there (data_dict[self.img_key]) is shaped (12,1,602,602). If i want to apply my transformation on the original input image (RGB) before going into feature dimensions where should i do it?

thank you!
Natalie.

**Comment by FabianIsensee** [comment]
Hey Natalie,
you already found the right place! The transforms are applied in the order that they are added to the tr_transforms list. The shape you are seeing is (b, c, x, y) where b is the batch size and c is the number of channels. Your dataset seems to only have one channel.
Before you proceed I suggest you update to master. We recently replaced the entire data augmentation pipeline. The new pipeline operates on samples, so the shape in your case will be (1, 602, 602)
Best,
Fabian

**Comment by natamendelson** [comment]
thank you for your reply,

the thing is that my input image is shaped (1,512,512). When it get to get_training_transforms the image size is already different. in data loader i have noticed the image might have been resampled - am i right? if so, i want to catch it even before. Our goal is to add stiches to the images randomly so it should be applied on the original image 512*512...

thank you,
Natalie.

**Comment by FabianIsensee** [comment]
Hey Natalie,
let me take a step back, I hope I am making all the right assumptions here:
- the images are 512x512
- your patch size is 512x512
- I don't believe your dataset requires and resampling

So you are wondering why the input to the data augmentation pipeline is 602x602? This is related to the SpatialTransform which rotates and rescales the image as data augmentation. If we were to do this on the intended patch size we would get a lot of black border artifacts where the rotation causes the image to rotate out of the patch area. We are trying to reduce this problem by first sampling larger patches (602x602 in your case) and then sampling the rotated 512x512 patch from that (center crop). In your particular case this extra step is not needed because we can only achieve 602x602 images by padding -> the result is the same as if we didn't do all this. But for other datasets where the patch size is smaller than the image size this is very useful!

If you want to apply your augmentation to a 512x512 image then apply it after the SpatialTransform.

I am a bit confused about the part where you ask about stitching. The dataloaders sample random patches from random locations. This is why they are not used in inference. Have a look at the code there to see how this is done.

Best,
Fabian
